packages/core/src/lib/passcode.test.ts

```diff
@@ -0,0 +1,77 @@
+import { PasscodeType } from '@logto/schemas';
+
+import {
+  deletePasscodeById,
+  findUnconsumedPasscodeBySessionIdAndType,
+  insertPasscode,
+} from '@/queries/passcode';
+
+import { createPasscode } from './passcode';
+
+jest.mock('@/queries/passcode');
+
+const mockedFindUnconsumedPasscodeBySessionIdAndType =
+  findUnconsumedPasscodeBySessionIdAndType as jest.MockedFunction<
+    typeof findUnconsumedPasscodeBySessionIdAndType
+  >;
+const mockedDeletePasscodeById = deletePasscodeById as jest.MockedFunction<
+  typeof deletePasscodeById
+>;
+const mockedInsertPasscode = insertPasscode as jest.MockedFunction<typeof insertPasscode>;
+
+beforeAll(() => {
+  mockedInsertPasscode.mockImplementation(async (data) => ({
+    ...data,
+    createdAt: Date.now(),
+    phone: data.phone ?? '',
+    email: data.email ?? '',
```

Comment:
no need for `??` since it's nullable?

Reply:
`data.phone` can be undefined, but the resolved `phone` cannot be undefined, so `??` is needed

Reply:
then should be `?? null`?